File: leetcode/Ruby/No16.3sum_closest.rb
=begin
Difficulty:
Medium

Desc:
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target.
Return the sum of the three integers.
You may assume that each input would have exactly one solution.

Example:
Given array nums = [-1, 2, 1, -4], and target = 1.
The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
=end

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
  nums.sort!
  result = nums[0] + nums[1] + nums[2]
  return result unless result != target

  nums.each_with_index { |num, index|
    i = index + 1
    j = nums.size - 1
    while i < j do
      sum = num + nums[i] + nums[j]
      if (target - sum).abs < (target - result).abs then
        result = sum
      end
      if sum > target then
        j -= 1
        while j > i && nums[j] == nums[j + 1] do
          j -= 1
        end
      elsif sum < target then
        i += 1
        while j > i && nums[i] == nums[i - 1] do
          i += 1
        end
      else
        return result
      end
    end
  }
  return result
end

# Test case
puts three_sum_closest([-1, 2, 1, -4], 1)
puts three_sum_closest([-1, 2, 1, -4], 2)
puts three_sum_closest([-1, 2, 1, -4], -3)

File: leetcode/C#/No16.3sum-closest.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target.
 * Return the sum of the three integers. You may assume that each input would have exactly one solution.
 *
 * Example 1:
 * Input: nums = [-1,2,1,-4], target = 1
 * Output: 2
 * Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
 *
 * Constraints:
 * 3 <= nums.length <= 10^3
 * -10^3 <= nums[i] <= 10^3
 * -10^4 <= target <= 10^4
*/

public class Solution {
    private int Search(int[] nums, int start, int target) {
        int i = start;
        int j = nums.Length - 1;

        int res = nums[i] + nums[j];
        while (i < j) {
            int num = nums[i] + nums[j];
            if (Math.Abs(num - target) < Math.Abs(res - target)) res = num;
            if (num == target) break;

            if (num < target) {
                i += 1;
            } else {
                j -= 1;
            }
        }
        return res;
    }

    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int res = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < nums.Length - 2; i += 1) {
            int sum = nums[i] + Search(nums, i + 1, target - nums[i]);
            if (Math.Abs(sum - target) < Math.Abs(res - target)) res = sum;
        }
        return res;
    }
}
